File: apps/libs/vending/softactivate/src/Licensing/ecc.cpp
//
// Copyright (c) 2014 FreeGrabApp. All rights reserved.
//

#include "precomp.h"
#include "ecc.h"
#include "bigint.h"
#include "ec2m.h"
#include "sha1.h"
#include "except.h"
#include "base64.h"
#include "uniconv.h"

#ifndef _WIN32
#include <alloca.h>
#endif

namespace ECC {

#if 0
void biPrint( bigint & data )
{
    int n;
    unsigned * u;

    n = ( data.bits() + 31 ) / 32;

    u = new unsigned[ n ];

    data.store( u, n );

    printf( "%d ", data.bits() );

    for ( int i = n - 1; i >= 0; i-- )
    {
        printf( "%08X ", u[i] );
    }

    printf("\n");

    delete [] u;
}
#endif

static const curve_parameter eccCurveTable[ 10 ] = { { 9, 7, 1, 17, 43, 48, 1},
                                              { 12, 7, 1, 83, 2049, 10, 1 }, 
                                              { 13, 7, 1, 27, 124, 52, 1},
                                              { 15, 7, 1, 3, 977, 110, 1},
                                              { 16, 7, 1, 45, 2902, 6, 1},
                                              { 10, 11, 2, 9, 139, 34, 1},
                                              { 12, 11, 2, 83, 2083, 66, 1},
                                              { 13, 11, 2, 27, 773, 48, 1},
                                              { 14, 11, 2, 43, 146, 40, 1},
                                              { 10, 17, 3, 9, 48, 28, 1} };

static void UintArrayToByteArray(unsigned * srcArray, int bitCount, unsigned char * dstArray)
{
    int dstLength = ((int)bitCount + 7) >> 3;
    int srcLength = (dstLength + 3) >> 2;

    int i = 0;
    int j = 0;
    unsigned currentUint;

    if (dstLength > 3)
    {
        while (i < dstLength - 3)
        {
            currentUint = (j < srcLength) ? srcArray[j] : 0;

            dstArray[i + 3] = (unsigned char)(currentUint >> 24);
            dstArray[i + 2] = (unsigned char)((currentUint >> 16) & 0xFF);
            dstArray[i + 1] = (unsigned char)((currentUint >> 8) & 0xFF);
            dstArray[i] = (unsigned char)(currentUint & 0xFF);
            i += 4;
            j += 1;
        }
    }

    currentUint = (j < srcLength) ? srcArray[j] : 0;
    if (i < dstLength) dstArray[i] = (unsigned char)(currentUint & 0xFF); i++;
    if (i < dstLength) dstArray[i] = (unsigned char)((currentUint >> 8) & 0xFF); i++;
    if (i < dstLength) dstArray[i] = (unsigned char)((currentUint >> 16) & 0xFF);
}

static void ByteArrayToUintArray(unsigned char * srcArray, int srcLength, unsigned * dstArray)
{
    int roundedLen = (srcLength + sizeof(unsigned) - 1) / sizeof(unsigned);

    int i = 0;
    int j = 0;
    while (i < roundedLen - 1)
    {
        dstArray[i] = (unsigned)(((unsigned)srcArray[j + 3] << 24) | ((unsigned)srcArray[j + 2] << 16) | ((unsigned)srcArray[j + 1] << 8) | (unsigned)srcArray[j]);
        i += 1;
        j += 4;
    }

    dstArray[roundedLen - 1] = 0;

    if (j < srcLength) dstArray[roundedLen - 1] = (unsigned)(srcArray[j]); j++;
    if (j < srcLength) dstArray[roundedLen - 1] |= (unsigned)((unsigned)srcArray[j] << 8); j++;
    if (j < srcLength) dstArray[roundedLen - 1] |= (unsigned)((unsigned)srcArray[j] << 16); j++;
    if (j < srcLength) dstArray[roundedLen - 1] |= (unsigned)((unsigned)srcArray[j] << 24);
}

class KeyImpl {
friend class Key;
friend class SignerImpl;
friend class VerifierImpl;

friend void GenerateKeyPair(KEY_TYPE keyType, Key **privKey, Key **pubKey);

private:
    KeyImpl()
    {

    }

    KeyImpl(const void * keyBuf, int bufLen, KEY_TYPE keyType = ECC)
    {
        Load(keyBuf, bufLen, keyType);      
    }

    KeyImpl(const char_t * base64Key, KEY_TYPE keyType = ECC)
    {
        USES_CONVERSION;
        BASE64 base64;

        int len;
        unsigned char * buf = base64.decode(T2A(base64Key), t_strlen(base64Key), &len);
        Load(buf, len, keyType);
        delete buf;
    }

    void Load(const void * keyBuf, int bufLen, KEY_TYPE keyType = ECC)
    {
        if (keyType == ECC)
        {
            unsigned char * keyBytes = (unsigned char *)keyBuf;

            if ((keyBytes[0] >> 5) != 0)
                throw new KeyExceptionImpl(STATUS_INVALID_PARAM, _T("unsupported key version"));

            if (keyBytes[1] + 2 != bufLen)
                throw new KeyExceptionImpl(STATUS_INVALID_PARAM, _T("invalid key size"));

            isPrivate = ((keyBytes[0] & 0x10) != 0);
            SetKeyType((KEY_TYPE)(keyBytes[0] & 0x0F));

            int roundedLen = sizeof(unsigned) * ((bufLen - 2 + sizeof(unsigned) - 1) / sizeof(unsigned));
            unsigned * buf = (unsigned *)calloc(roundedLen, 1);
            if (!buf) return;

            ByteArrayToUintArray(keyBytes + 2, bufLen - 2, buf);

            m_key.load(buf, (unsigned)(roundedLen / sizeof(unsigned)));
        } else
        {
            SetKeyType(keyType);

            int roundedLen = sizeof(unsigned) * ((bufLen + sizeof(unsigned) - 1) / sizeof(unsigned));
            unsigned * buf = (unsigned *)calloc(roundedLen, 1);
            if (!buf) return;

            ByteArrayToUintArray((unsigned char *)keyBuf, bufLen, buf);

            m_key.load(buf, (unsigned)(roundedLen / sizeof(unsigned)));
        }
    }

    bool Store(void * buf, int * bufLen)
    {
        unsigned char * keyBytes = (unsigned char *)buf;

        int numBits = m_key.bits();

        int numBytes = (numBits + 7) >> 3;

        if (*bufLen < 2 + numBytes)
            return false;

        keyBytes[0] &= 0x1F; // set first 3 bits to 0 - version 1
        if (isPrivate) keyBytes[0] |= 0x10; else keyBytes[0] &= 0xEF; // the 4th bit is 1 if it's a private key
        keyBytes[0] = (keyBytes[0] & 0xF0) | (((unsigned char)m_keyType) & 0x0F); // the next 4 bits from 1st byte are the key type
        keyBytes[1] = numBytes; // the second byte specifies the key size

        int numUints = (numBytes + sizeof(unsigned) - 1) / sizeof(unsigned);
        unsigned * temp = (unsigned int *)_alloca(numUints * sizeof(unsigned));
        m_key.store(temp, numUints);

        UintArrayToByteArray(temp, numBits, keyBytes + 2);

        *bufLen = numBytes + 2;

        return true;
    }

    void SetKeyType(KEY_TYPE keyType)
    {
        m_keyType = keyType;
        switch (keyType)
        {
        case ECC_54: m_keySize = 54;break;
        case ECC_73: m_keySize = 73;break;
        case ECC_79: m_keySize = 79;break;
        case ECC_91: m_keySize = 91;break;
        case ECC_97: m_keySize = 97;break;
        case ECC_100: m_keySize = 100;break;
        case ECC_120: m_keySize = 120;break;
        case ECC_131: m_keySize = 131;break;
        case ECC_141: m_keySize = 141;break;
        case ECC_161: m_keySize = 161;break;
        }
    }

    KEY_TYPE GetKeyType()
    {
        return m_keyType;
    }

    KEY_TYPE m_keyType;
    int m_keySize;
    bigint m_key;
    bool isPrivate;
};

class SignerImpl {
public:
    SignerImpl()
    {
        m_hashBits = 0; // thus select ECDSA as the signing algorithm
    }

    ~SignerImpl()
    {

    }

    static bigint Hash(const bigint & x, const bigint & p)
    {
        unsigned int n = ( x.bits() + ( ( sizeof(unsigned) << 3 ) - 1 ) ) / ( sizeof( unsigned ) << 3 );
        unsigned int * a = (unsigned *)_alloca(n * sizeof(unsigned));
        unsigned int h[ 5 ];

        x.store(a, n);

        SHA1 sha;
        sha.Update((unsigned char *)a, n * sizeof(unsigned));
        sha.Final();
        sha.GetHash((unsigned char *)h);

        bigint r;
        r.load( h, 5 );

        return r % p;
    }

    void SetHashSize(int hashBits)
    {
        m_hashBits = hashBits;
    }

    void SetPrivateKey(const KeyImpl & privKey)
    {
        m_privateKey = privKey;
    }

    void Sign(const void * msg, int msgLen, void * sigBuf, int * sigLen, int * sigLenBits)
    {
        full_curve_parameter param(eccCurveTable[ m_privateKey.m_keyType ]);
        ECC_BASE ecc(param);
        bigint message;
        bigint temp;
        pair signature;

        *sigLenBits = (m_hashBits && m_hashBits < m_privateKey.m_keySize) ? m_hashBits + m_privateKey.m_keySize : m_privateKey.m_keySize << 1;
        *sigLen = (*sigLenBits + 7) >> 3;

        unsigned * buf = (unsigned *)_alloca(msgLen + *sigLen + sizeof(unsigned) - 1);
        // take care of endianess here, do not do just a memcpy()
        memcpy(buf, msg, msgLen);
        memset((char *)buf + msgLen, 0, sizeof(unsigned) - 1);
        
        unsigned msgHash[5];

        SHA1 sha;
        sha.Update((unsigned char *)msg, msgLen);
        sha.Final();

        sha.GetHash((unsigned char *)msgHash);

        message.load(msgHash, 5);

        signature = (m_hashBits && m_hashBits < m_privateKey.m_keySize) ? ecc.schnorr_sign(message, m_privateKey.m_key, Hash, m_hashBits) : ecc.dsa_sign(message, m_privateKey.m_key, Hash);

        temp = (signature.r << m_privateKey.m_keySize); temp += signature.s;

        temp.store(buf, (unsigned)((*sigLen + sizeof(unsigned) - 1) / sizeof(unsigned)));
         
        // take care of endianess here
        memcpy(sigBuf, buf, *sigLen);
    }

private:
    KeyImpl m_privateKey;
    int m_hashBits;
};

class VerifierImpl {
public:
    VerifierImpl()
    {
        m_hashBits = 0;
    }

    ~VerifierImpl()
    {

    }

    static bigint Hash(const bigint & x, const bigint & p)
    {
        int n = ( x.bits() + ( ( sizeof(unsigned) << 3 ) - 1 ) ) / ( sizeof( unsigned ) << 3 );
        unsigned int * a = (unsigned *)_alloca(n * sizeof(unsigned));
        unsigned int h[ 5 ];

        x.store(a, n);

        SHA1 sha;
        sha.Update((unsigned char *)a, (x.bits() + 7) >> 3);
        sha.Final();
        sha.GetHash((unsigned char *)h);

        bigint r;
        r.load( h, 5 );

        return r % p;
    }

    void SetHashSize(int hashBits)
    {
        m_hashBits = hashBits;
    }

    void SetPublicKey(const KeyImpl & pubKey)
    {
        m_publicKey = pubKey;
    }

    bool Verify(const void * msg, int msgLen, const void * sig, int sigLen, int sigLenBits)
    {
        full_curve_parameter param(eccCurveTable[ m_publicKey.m_keyType ]);
        ECC_BASE ecc(param);
        bigint message;
        pair signature;

        int msgBufLen = (msgLen +  sizeof(unsigned) - 1) / sizeof(unsigned);
        int sigBufLen = (sigLen +  sizeof(unsigned) - 1) / sizeof(unsigned);

        unsigned * msgBuf = (unsigned *)_alloca(msgBufLen * sizeof(unsigned));
        unsigned * sigBuf = (unsigned *)_alloca(sigBufLen * sizeof(unsigned));
        memset(msgBuf, 0, msgBufLen * sizeof(unsigned));
        memcpy(msgBuf, msg, msgLen);
        memset(sigBuf, 0, sigBufLen * sizeof(unsigned));
        memcpy(sigBuf, sig, sigLen);

        unsigned int msgHash[5];

        SHA1 sha;
        sha.Update((unsigned char *)msg, msgLen);
        sha.Final();

        sha.GetHash((unsigned char *)msgHash);

        message.load(msgHash, 5);

        signature.r.load(sigBuf, sigBufLen); signature.r >>= m_publicKey.m_keySize; signature.r &= (pow2((unsigned)(sigLenBits - m_publicKey.m_keySize)) - 1);
        signature.s.load(sigBuf, sigBufLen); signature.s &= (pow2(m_publicKey.m_keySize) - 1);

        return (m_hashBits && m_hashBits < m_publicKey.m_keySize) ? ecc.schnorr_verify(message, signature, m_publicKey.m_key, Hash, m_hashBits) : ecc.dsa_verify(message, signature, m_publicKey.m_key, Hash);
    }

private:
    int m_hashBits;
    KeyImpl m_publicKey;
};

void GenerateKeyPair(KEY_TYPE keyType, Key **privKey, Key **pubKey)
{
    full_curve_parameter param(eccCurveTable[ keyType ]);
    ECC_BASE ecc(param);
    Key * privateKey = new Key(), 
        * publicKey = new Key();

    privateKey->m_Impl.m_key = ecc.create_private_key();
    privateKey->m_Impl.SetKeyType(keyType);
    privateKey->m_Impl.isPrivate = true;

    publicKey->m_Impl.m_key = ecc.create_public_key(privateKey->m_Impl.m_key);
    publicKey->m_Impl.SetKeyType(keyType);

    *privKey = privateKey;
    *pubKey = publicKey;
}

Key::Key():
m_Impl( *new KeyImpl() )
{

}

Key::Key(const void * keyBuf, int bufLen, KEY_TYPE keyType):
m_Impl( *new KeyImpl(keyBuf, bufLen, keyType) )
{

}

Key::Key(const char_t * base64Key, KEY_TYPE keyType):
m_Impl( *new KeyImpl(base64Key, keyType))
{

}

Key::~Key()
{
    delete & m_Impl;
}

void Key::Load(const void * keyData, int keyLen, KEY_TYPE keyType)
{
    m_Impl.Load(keyData, keyLen, keyType);
}

bool Key::Store(void * buf, int * bufLen) const
{
    return m_Impl.Store(buf, bufLen);
}

const char_t * Key::ToString()
{
    USES_CONVERSION;
    static string_t base64key;

    int len = 2 + ((m_Impl.m_key.bits() + 7) >> 3);
    unsigned char * buf = (unsigned char *)_alloca(len);
    Store(buf, &len);

    BASE64 base64;

    char * base64Buf = base64.encode(buf, len, true);
    if (!base64Buf)
        throw new KeyExceptionImpl(STATUS_GENERIC_ERROR, _T("could not encode buffer to base64"));

    base64key = A2T(base64Buf);
    free(base64Buf);

    return base64key.c_str();
}

KEY_TYPE Key::GetKeyType()
{
    return m_Impl.GetKeyType();
}

const Key & Key::operator = (const Key & key)
{
    m_Impl = key.m_Impl;
    return (const Key &)(*this);
}

Signer::Signer():
m_Impl(*new SignerImpl())
{

}

Signer::~Signer()
{
    delete & m_Impl;
}

void Signer::SetHashSize(int hashBits)
{
    m_Impl.SetHashSize(hashBits);
}

void Signer::SetPrivateKey(const Key * privKey)
{
    m_Impl.SetPrivateKey(privKey->m_Impl);
}

void Signer::Sign(const void * msg, int msgLen, void * sigBuf, int * sigLen, int * sigLenBits)
{
    m_Impl.Sign(msg, msgLen, sigBuf, sigLen, sigLenBits);
}

Verifier::Verifier():
m_Impl( *new VerifierImpl() )
{

}

Verifier::~Verifier()
{
    delete & m_Impl;
}

void Verifier::SetHashSize(int hashBits)
{
    m_Impl.SetHashSize(hashBits);
}

void Verifier::SetPublicKey(const Key * pubKey)
{
    m_Impl.SetPublicKey(pubKey->m_Impl);
}

bool Verifier::Verify(const void * msg, int msgLen, const void * sig, int sigLen, int sigLenBits)
{
    return m_Impl.Verify(msg, msgLen, sig, sigLen, sigLenBits);
}

}

File: apps/libs/vending/softactivate/src/Licensing.Net/ECC.cs
//
// Copyright (c) 2014 SoftActivate. All rights reserved.
//

using System;
using System.Collections.Generic;
using System.Text;
using System.Security.Cryptography;

namespace SoftActivate.Licensing
{
    sealed class Ecc
    {
        internal enum KeyType
        {
            Ecc54,
            Ecc73,
            Ecc79,
            Ecc91,
            Ecc97,
            Ecc100,
            Ecc120,
            Ecc131,
            Ecc141,
            Ecc161,
            Ecc,
            EccCustom,
        }

        Ecc()
        {

        }

        public static void GenerateKeyPair(KeyType keyType, out Key privKey, out Key pubKey)
        {
            EccBase ecc = new EccBase((int)keyType);
            privKey = new Key();
            privKey.KeyType = keyType;
            privKey.BigIntKey     = ecc.CreatePrivateKey();
            pubKey = new Key();
            pubKey.KeyType  = keyType;
            pubKey.BigIntKey  = ecc.CreatePublicKey(privKey.BigIntKey);
        }

        public static void UintArrayToByteArray(uint[] srcArray, uint bitCount, out byte[] dstArray)
        {
            int dstLength = ((int)bitCount + 7) >> 3;
            dstArray = new byte[dstLength];

            int i = 0;
            int j = 0;
            uint currentUint;

            if (dstLength > 3)
            {
                while (i < dstLength - 3)
                {
                    currentUint = (j < srcArray.Length) ? srcArray[j] : 0;

                    dstArray[i + 3] = (byte)(currentUint >> 24);
                    dstArray[i + 2] = (byte)((currentUint >> 16) & 0xFF);
                    dstArray[i + 1] = (byte)((currentUint >> 8) & 0xFF);
                    dstArray[i] = (byte)(currentUint & 0xFF);
                    i += 4;
                    j += 1;
                }
            }

            currentUint = (j < srcArray.Length) ? srcArray[j] : 0;
            if (i < dstLength) dstArray[i] = (byte)(currentUint & 0xFF); i++;
            if (i < dstLength) dstArray[i] = (byte)((currentUint >> 8) & 0xFF); i++;
            if (i < dstLength) dstArray[i] = (byte)((currentUint >> 16) & 0xFF);
        }

        public static void ByteArrayToUintArray(byte[] srcArray, int offset, int count, out uint[] dstArray)
        {
            int roundedLen = (count + sizeof(uint) - 1) / sizeof(uint);
            dstArray = new uint[ roundedLen ];

            int i = 0;
            int j = offset;
            while (i < roundedLen - 1)
            {
                dstArray[i] = (uint)((srcArray[j + 3] << 24) | (srcArray[j + 2] << 16) | (srcArray[j + 1] << 8) | srcArray[j]);
                i += 1;
                j += 4;
            }

            dstArray[roundedLen - 1] = 0;

            int srcEndOffset = offset + count;

            if (j < srcEndOffset) dstArray[roundedLen - 1] = (uint)(srcArray[j]); j++;
            if (j < srcEndOffset) dstArray[roundedLen - 1] |= (uint)(srcArray[j] << 8); j++;
            if (j < srcEndOffset) dstArray[roundedLen - 1] |= (uint)(srcArray[j] << 16); j++;
            if (j < srcEndOffset) dstArray[roundedLen - 1] |= (uint)(srcArray[j] << 24);
        }
    }

    class Key
    {
    	public Key()
        {

        }

        public Key(string base64key, Ecc.KeyType keyType = Ecc.KeyType.Ecc)
        {
            Load(Convert.FromBase64String(base64key), keyType);
        }

    	public Key(byte[] keyData)
        {
            Load(keyData);
        }

	    public void Load(byte[] keyData, Ecc.KeyType keyType = Ecc.KeyType.Ecc)
        {
            if (keyType == Ecc.KeyType.Ecc)
            {
                if ((keyData[0] >> 5) != 0)
                    throw new Exception("unsupported key version");

                if (2 + keyData[1] != keyData.Length)
                    throw new Exception("invalid embedded key length");

                isPrivate = ((keyData[0] & 0x10) != 0);

                uint[] buf;

                Ecc.ByteArrayToUintArray(keyData, 2, keyData.Length - 2, out buf);

                key = new BigInt();
                key.Load(buf, (uint)buf.Length);

                SetKeyType((Ecc.KeyType)(keyData[0] & 0x0F));
            }
            else
            {
                uint[] buf;

                Ecc.ByteArrayToUintArray(keyData, 0, keyData.Length, out buf);

                key = new BigInt();
                key.Load(buf, (uint)buf.Length);

                SetKeyType(keyType);
            }
        }

	    public byte[] Export()
        {
            uint[] temp;
            byte[] keyBuf;

            key.Store(out temp);
            Ecc.UintArrayToByteArray(temp, key.Bits, out keyBuf);

            byte[] buf = new byte[2 + keyBuf.Length];

            buf[0] = (byte)(((byte)0 << 5) | ((byte)keyType & 0x0F));

            if (isPrivate) buf[0] |= (0x10);

            buf[1] = (byte)keyBuf.Length;

            Buffer.BlockCopy(keyBuf, 0, buf, 2, keyBuf.Length);

            return buf;
        }
        
        private void SetKeyType(Ecc.KeyType type)
        {
            keyType = type;
            switch (keyType)
            {
                case Ecc.KeyType.Ecc54: keySize = 54; break;
                case Ecc.KeyType.Ecc73: keySize = 73; break;
                case Ecc.KeyType.Ecc79: keySize = 79; break;
                case Ecc.KeyType.Ecc91: keySize = 91; break;
                case Ecc.KeyType.Ecc97: keySize = 97; break;
                case Ecc.KeyType.Ecc100: keySize = 100; break;
                case Ecc.KeyType.Ecc120: keySize = 120; break;
                case Ecc.KeyType.Ecc131: keySize = 131; break;
                case Ecc.KeyType.Ecc141: keySize = 141; break;
                case Ecc.KeyType.Ecc161: keySize = 161; break;
            }
        }

        public uint KeySize
        {
            get
            {
                return keySize;
            }
        }

        public BigInt BigIntKey
        {
            get
            {
                return key;
            }
            set
            {
                key = value;
            }
        }

        public Ecc.KeyType KeyType
        {
            get
            {
                return keyType;
            }
            set
            {
                SetKeyType(value);
            }
        }

        public bool IsPrivate
        {
            get
            {
                return isPrivate;
            }
            set
            {
                isPrivate = value;
            }
        }

        private Ecc.KeyType keyType;
        private uint keySize;
        private BigInt key;
        private bool isPrivate;
    }

    delegate uint HashCallback(uint[] data, uint dataBits, uint[] hashedData, uint hashBits);

    class Signer
    {
	    public Signer()
        {

        }

	    ~Signer()
        {

        }

	    public void SetPrivateKey(Key privKey)
        {
            privateKey = privKey;
        }

	    public void SetHashSize(int hashBits)
        {
            this.hashBits = (uint)hashBits;
        }

	    BigInt Hash(BigInt x, BigInt p)
	    {
            uint[] uintX;
            x.Store(out uintX);

            byte[] byteX;
            Ecc.UintArrayToByteArray(uintX, x.Bits, out byteX);

            byte[] hash = sha1.ComputeHash(byteX);
            
            uint[] uintHash;
            Ecc.ByteArrayToUintArray(hash, 0, hash.Length, out uintHash);

            BigInt h = new BigInt();
            h.Load(uintHash, (uint)uintHash.Length);

            return h % p;
        }

        public void Sign(byte[] msg, int offset, int count, out byte[] sig, out int sigBits)
        {
            EccBase ecc = new EccBase((int)privateKey.KeyType);
		    BigInt message = new BigInt();
		    BigInt temp;
            Pair signature = new Pair();

            msg = sha1.ComputeHash(msg, offset, count);

            sigBits = (int)((hashBits > 0 && hashBits < privateKey.KeySize) ? hashBits + privateKey.KeySize : privateKey.KeySize << 1);

            uint[] buf;
            Ecc.ByteArrayToUintArray(msg, 0, msg.Length, out buf);

            message.Load(buf, (uint)buf.Length);

            //System.Diagnostics.Debug.WriteLine("[sign] Private Key: " + privateKey.Key.ToString());
            //System.Diagnostics.Debug.WriteLine("[sign] Message: " + message.ToString());

		    signature = (hashBits > 0 && hashBits < privateKey.KeySize) ? ecc.SchnorrSign(message, privateKey.BigIntKey, Hash, hashBits) : ecc.DsaSign(message, privateKey.BigIntKey, Hash);

            //System.Diagnostics.Debug.WriteLine("[sign] Signature.R: " + signature.r.ToString());
            //System.Diagnostics.Debug.WriteLine("[sign] Signature.S: " + signature.s.ToString());

		    temp = (signature.r << (int)privateKey.KeySize); temp += signature.s;

            temp.Store(out buf);

            Ecc.UintArrayToByteArray(buf, (uint)sigBits, out sig); 
        }

        private Key privateKey;
        private uint hashBits;
        private SHA1CryptoServiceProvider sha1 = new SHA1CryptoServiceProvider();
    }

    class Verifier 
    {
    	public Verifier()
        {

        }

    	~Verifier()
        {

        }

        BigInt Hash(BigInt x, BigInt p)
        {
            uint[] uintX;
            x.Store(out uintX);

            byte[] byteX;
            Ecc.UintArrayToByteArray(uintX, x.Bits, out byteX);

            byte[] hash = sha1.ComputeHash(byteX);

            uint[] uintHash;
            Ecc.ByteArrayToUintArray(hash, 0, hash.Length, out uintHash);

            BigInt h = new BigInt();
            h.Load(uintHash, (uint)uintHash.Length);

            return h % p;
        }

    	public void SetPublicKey(Key publicKey)
        {
            this.publicKey = publicKey;
        }

    	public void SetHashSize(uint hashBits)
        {
            this.hashBits = hashBits;
        }

        public bool Verify(byte[] msg, int msgOffset, int msgCount, byte[] sig, int sigOffset, int sigCount, int sigLenBits)
        {
            EccBase ecc = new EccBase((int)publicKey.KeyType);
		    BigInt message = new BigInt();
		    Pair signature = new Pair();

            msg = sha1.ComputeHash(msg, msgOffset, msgCount);

            uint[] msgBuf;
            Ecc.ByteArrayToUintArray(msg, 0, msg.Length, out msgBuf);

            uint[] sigBuf;
            Ecc.ByteArrayToUintArray(sig, sigOffset, sigCount, out sigBuf);

		    message.Load(msgBuf, (uint)msgBuf.Length);
		    signature.r.Load(sigBuf, (uint)sigBuf.Length); signature.r = signature.r >> (int)publicKey.KeySize; signature.r &= (BigInt.Pow2((uint)(sigLenBits - publicKey.KeySize)) - 1);
		    signature.s.Load(sigBuf, (uint)sigBuf.Length); signature.s &= (BigInt.Pow2(publicKey.KeySize) - 1);

            //System.Diagnostics.Debug.WriteLine("[verify] Public Key: " + publicKey.Key.ToString());
            //System.Diagnostics.Debug.WriteLine("[verify] Message: " + message.ToString());

            //System.Diagnostics.Debug.WriteLine("[verify] Signature.R: " + signature.r.ToString());
            //System.Diagnostics.Debug.WriteLine("[verify] Signature.S: " + signature.s.ToString());

		    return (hashBits != 0 && hashBits < publicKey.KeySize) ? ecc.SchnorrVerify(message, signature, publicKey.BigIntKey, Hash, hashBits) : ecc.DsaVerify(message, signature, publicKey.BigIntKey, Hash);
        }

        private uint hashBits;
        private Key publicKey;
        private SHA1CryptoServiceProvider sha1 = new SHA1CryptoServiceProvider();
    }
}
